control-plane-operator/hostedclusterconfigoperator/controllers/resources/globalps/pullsecret.go

```diff
@@ -0,0 +1,297 @@
+	var (
+		originalAuths         map[string]any
+		userProvidedAuths     map[string]any
+		originalJSON          map[string]any
+		userProvidedJSON      map[string]any
+		globalPullSecretBytes []byte
+		err                   error
+	)
+
+	// Unmarshal original pull secret
+	if err = json.Unmarshal(originalPullSecret, &originalJSON); err != nil {
+		return nil, fmt.Errorf("invalid original pull secret format: %w", err)
+	}
+	originalAuths = originalJSON["auths"].(map[string]any)
+
+	// Unmarshal additional pull secret
+	if err = json.Unmarshal(userProvidedPullSecret, &userProvidedJSON); err != nil {
+		return nil, fmt.Errorf("invalid user provided pull secret format: %w", err)
+	}
+	userProvidedAuths = userProvidedJSON["auths"].(map[string]any)
+
+	// Merge auths
+	for k, v := range userProvidedAuths {
+		originalAuths[k] = v
+	}
+
+	// Create final JSON
+	finalJSON := map[string]any{
+		"auths": originalAuths,
+	}
+
+	globalPullSecretBytes, err = json.Marshal(finalJSON)
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal merged pull secret: %w", err)
+	}
+
+	return globalPullSecretBytes, nil
+}
+
+func ReconcileGlobalPullSecretRBAC(ctx context.Context, c crclient.Client, createOrUpdate upsert.CreateOrUpdateFN, namespace string) error {
+	// Remove the RBAC resources if the user provided pull secret is not present
+	log, err := logr.FromContext(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get logger: %w", err)
+	}
+
+	exists, err := UserProvidedPullSecretExists(c, ctx)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to check if user provided pull secret exists: %w", err)
+		}
+	}
+
+	log.Info("Reconciling global pull secret RBAC", "exists", exists)
+
+	if !exists {
+		log.Info("Deleting global pull secret RBAC resources", "exists", exists)
+		sa := manifests.GlobalPullSecretServiceAccount()
+		if err := c.Delete(ctx, sa); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return fmt.Errorf("failed to delete service account for global pull secret: %w", err)
+			}
+		}
+		clusterRole := manifests.GlobalPullSecretClusterRole()
+		if err := c.Delete(ctx, clusterRole); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return fmt.Errorf("failed to delete cluster role for global pull secret: %w", err)
+			}
+		}
+		clusterRoleBinding := manifests.GlobalPullSecretClusterRoleBinding()
+		if err := c.Delete(ctx, clusterRoleBinding); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return fmt.Errorf("failed to delete cluster role binding for global pull secret: %w", err)
+			}
+		}
+
+		return nil
+	}
+
+	log.Info("Creating global pull secret RBAC resources", "exists", exists)
+
+	// Create ServiceAccount
+	sa := manifests.GlobalPullSecretServiceAccount()
+	if _, err := createOrUpdate(ctx, c, sa, func() error { return nil }); err != nil {
+		return fmt.Errorf("failed to reconcile service account: %w", err)
+	}
+
+	// Create ClusterRole
+	clusterRole := manifests.GlobalPullSecretClusterRole()
```

Comment:
please make this namespaced roles / roles binding in the target namespaces they need to operate and use ResourceNames to scope them to the specific secrets